aea/runtime.py

```diff
@@ -0,0 +1,238 @@
+"""This module contains the implementation of runtime for economic agent (AEA)."""
+import asyncio
+import logging
+import threading
+from abc import ABC, abstractmethod
+from asyncio.events import AbstractEventLoop
+from enum import Enum
+from typing import Optional
+
+from aea.agent_loop import AsyncState
+from aea.helpers.async_utils import ensure_loop
+from aea.mail.base import AsyncMultiplexer
+
+if False:
+    # for mypy
+    from aea.agent import Agent
+
+
+logger = logging.getLogger(__name__)
+
+
+class RuntimeStates(Enum):
+    """Runtime states."""
+
+    initial = "not_started"
+    starting = "starting"
+    started = "started"
+    stopped = "stopped"
+
+
+class BaseRuntime(ABC):
+    """Abstract runtime class to create implementations."""
+
+    def __init__(
+        self, agent: "Agent", loop: Optional[AbstractEventLoop] = None
+    ) -> None:
+        """
+        Init runtime.
+
+        :param agent: Agent to run.
+        :param loop: optional event loop. if not provided a new one will be created.
+        :return: None
+        """
+        self._agent: "Agent" = agent
+        self._loop = ensure_loop(
+            loop
+        )  # TODO: decide who constructs loop: agent, runtime, multiplexer.
+        self._state: AsyncState = AsyncState(RuntimeStates.initial)
+
+    def start(self) -> None:
+        """Start agent using runtime."""
+        if self._state.get() is RuntimeStates.started:
+            logger.error("Already started!")
+            return
+        self._start()
+
+    def stop(self) -> None:
+        """Stop agent and runtime."""
+        logger.debug("Runtime stop called!")
+        self.teardown()
+        self._stop()
+        logger.debug("[{}]: Stopped".format(self._agent.name))
+
+    def teardown(self) -> None:
+        """Tear down agent."""
+        logger.debug("Runtime teardown ...")
+        self._agent.teardown()
+        logger.debug("[{}]: Teardown completed".format(self._agent.name))
+
+    @abstractmethod
+    def _start(self) -> None:
+        """Implement runtime start function here."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def _stop(self) -> None:
+        """Implement runtime stop function here."""
+        raise NotImplementedError
+
+
+class AsyncRuntime(BaseRuntime):
+    """Asynchronous runtime: uses asyncio loop for multiplexer and async agent main loop."""
+
+    def __init__(
+        self, agent: "Agent", loop: Optional[AbstractEventLoop] = None
+    ) -> None:
+        """
+        Init runtime.
+
+        :param agent: Agent to run.
+        :param loop: optional event loop. if not provided a new one will be created.
+        :return: None
+        """
+        super().__init__(agent=agent, loop=loop)
+        self._stopping_task: Optional[asyncio.Task] = None
+        self._async_stop_lock: Optional[asyncio.Lock] = None
+
+    def _start(self) -> None:
+        """Start runtime."""
```

Comment:
Provide some details on what this does in the docstring.

Reply:
updated